src/modules/frost/tests_impl.h

```diff
@@ -0,0 +1,738 @@
+
+#include <stdlib.h>
+#include <string.h>
+
+#include "../../../include/secp256k1.h"
+#include "../../../include/secp256k1_extrakeys.h"
+#include "../../../include/secp256k1_frost.h"
+
+#include "session.h"
+#include "keygen.h"
+#include "../../scalar.h"
+#include "../../scratch.h"
+#include "../../field.h"
+#include "../../group.h"
+#include "../../hash.h"
+#include "../../util.h"
+
+/* Simple (non-adaptor, non-tweaked) 3-of-5 FROST aggregate, sign, verify
+ * test. */
+void frost_simple_test(void) {
+    secp256k1_frost_pubnonce pubnonce[5];
+    const secp256k1_frost_pubnonce *pubnonce_ptr[5];
+    unsigned char msg[32];
+    secp256k1_xonly_pubkey pk;
+    unsigned char seed[32];
+    secp256k1_frost_share shares[5];
+    secp256k1_frost_secnonce secnonce[5];
+    secp256k1_pubkey pubshares[5];
+    secp256k1_frost_partial_sig partial_sig[5];
+    const secp256k1_frost_partial_sig *partial_sig_ptr[5];
+    unsigned char final_sig[64];
+    secp256k1_frost_session session;
+    int i;
+    size_t ids[5];
+
+    secp256k1_testrand256(seed);
+
+    for (i = 0; i < 5; i++) {
+        pubnonce_ptr[i] = &pubnonce[i];
+        partial_sig_ptr[i] = &partial_sig[i];
+        ids[i] = i + 1;
+    }
+
+    CHECK(secp256k1_frost_shares_trusted_gen(CTX, shares, pubshares, &pk, seed, 3, 5) == 1);
+
+    secp256k1_testrand256(msg);
+    for (i = 0; i < 3; i++) {
+        unsigned char session_id[32];
+
+        secp256k1_testrand256(session_id);
+
+        CHECK(secp256k1_frost_nonce_gen(CTX, &secnonce[i], &pubnonce[i], session_id, &shares[i], NULL, NULL, NULL) == 1);
+    }
+    for (i = 0; i < 3; i++) {
+        CHECK(secp256k1_frost_nonce_process(CTX, &session, pubnonce_ptr, 3, msg, &pk, ids[i], ids, NULL, NULL) == 1);
+        CHECK(secp256k1_frost_partial_sign(CTX, &partial_sig[i], &secnonce[i], &shares[i], &session, NULL) == 1);
+        CHECK(secp256k1_frost_partial_sig_verify(CTX, &partial_sig[i], &pubnonce[i], &pubshares[i], &session, NULL) == 1);
+    }
+    CHECK(secp256k1_frost_partial_sig_agg(CTX, final_sig, &session, partial_sig_ptr, 3) == 1);
+    CHECK(secp256k1_schnorrsig_verify(CTX, final_sig, msg, sizeof(msg), &pk) == 1);
+}
+
+void frost_pubnonce_summing_to_inf(secp256k1_frost_pubnonce *pubnonce) {
+    secp256k1_ge ge[2];
+    int i;
+    secp256k1_gej summed_nonces[2];
+    const secp256k1_frost_pubnonce *pubnonce_ptr[2];
+
+    ge[0] = secp256k1_ge_const_g;
+    ge[1] = secp256k1_ge_const_g;
+
+    for (i = 0; i < 2; i++) {
+        secp256k1_frost_pubnonce_save(&pubnonce[i], ge);
+        pubnonce_ptr[i] = &pubnonce[i];
+        secp256k1_ge_neg(&ge[0], &ge[0]);
+        secp256k1_ge_neg(&ge[1], &ge[1]);
+    }
+
+    secp256k1_frost_sum_nonces(CTX, summed_nonces, pubnonce_ptr, 2);
+    CHECK(secp256k1_gej_is_infinity(&summed_nonces[0]));
+    CHECK(secp256k1_gej_is_infinity(&summed_nonces[1]));
+}
+
+int frost_memcmp_and_randomize(unsigned char *value, const unsigned char *expected, size_t len) {
```

Comment:
The same function is also used in MuSig's tests, so we can make it a helper function in `src/testutil.h`.